File: medium/399. Evaluate Division/399. Evaluate Division.ts
function calcEquation(equations: string[][], values: number[], queries: string[][]): number[] {
    const graph: { [key in string]: { [key in string]: number } } = {}
    // Build graph
    for (let i = 0; i < equations.length; i++) {
        const [A, B] = equations[i]
        const value = values[i]

        graph[A] = { ...graph[A], [B]: value }
        graph[B] = { ...graph[B], [A]: 1 / value }
    }

    // Search function for calculating
    const dfs = (C: string, D: string, visited: Set<string>): number => {
        // Check if there is no variable in the whole graph
        if (!(C in graph) || !(D in graph)) {
            return -1.0
        }

        // If one variable is the same to another: A / A = 1
        if (C === D) {
            return 1.0
        }

        visited.add(C)

        for (const neighbor in graph[C]) {
            if (!visited.has(neighbor)) {
                const value = graph[C][neighbor]
                const result = dfs(neighbor, D, visited)
                if (result !== -1.0) {
                    return result * value
                }
            }
        }

        return -1.0
    }

    // Calculate queries
    const result: number[] = []

    for (const [C, D] of queries) {
        result.push(dfs(C, D, new Set()))
    }

    return result
};

console.log(calcEquation([["a", "b"], ["b", "c"]], [2.0, 3.0], [["a", "c"], ["b", "a"], ["a", "e"], ["a", "a"], ["x", "x"]]))

export { }

File: medium/399. Evaluate Division/399. Evaluate Division.py
from collections import defaultdict
from typing import List, Set


class Solution:
    def calcEquation(
        self, equations: List[List[str]], values: List[float], queries: List[List[str]]
    ) -> List[float]:
        graph = defaultdict(dict)
        print(f"{graph=}")

        # Build a graph
        for (A, B), value in zip(equations, values):
            graph[A][B] = value
            graph[B][A] = 1 / value

        # The DFS function to search the path
        def dfs(start: str, end: str, visited: Set):
            if start not in graph or end not in graph:
                return -1.0
            if start == end:
                return 1.0

            visited.add(start)

            for neighbor, weight in graph[start].items():
                if neighbor not in visited:
                    result = dfs(neighbor, end, visited)
                    if result != -1.0:
                        return result * weight

            return -1.0

        print(f"{graph=}")

        # Handle queries
        results = []
        for C, D in queries:
            results.append(dfs(C, D, set()))

        return results


solution = Solution()
print(
    solution.calcEquation(
        equations=[["a", "b"], ["b", "c"]],
        values=[2.0, 3.0],
        queries=[["a", "c"], ["b", "a"], ["a", "e"], ["a", "a"], ["x", "x"]],
    )
)
